File: backend/app/services/linux_services.py
"""Linux service discovery and status.

Resolution order:
1. systemd (only if PID 1 is really systemd; avoids the typical WSL error).
2. SysV via /etc/init.d.
3. If no manager is accessible, a controlled response with available=False.

`get_services_with_tracking` also resolves the status of configured
tracked services (nginx, docker, ...) including "not installed".
"""

import os
import shutil
import subprocess

from ..schemas.services import ServiceInfo, ServicesResponse, TrackedService

_STATE_LABELS = {
    "active": "Activo",
    "inactive": "Inactivo",
    "failed": "Fallido",
    "activating": "Activando",
    "deactivating": "Desactivando",
    "exited": "Finalizado",
    "reloading": "Recargando",
    "registered": "Registrado",
}


def _translate(state: str) -> str:
    return _STATE_LABELS.get(state, state.title() if state else "Desconocido")


def _systemctl_available() -> bool:
    return shutil.which("systemctl") is not None


def _system_is_booted_with_systemd() -> bool:
    """Checks whether PID 1 is systemd (avoids the typical WSL error)."""
    try:
        with open("/proc/1/comm", "r", encoding="utf-8") as proc:
            return proc.read().strip() == "systemd"
    except OSError:
        return False


def _safe_run(cmd: list[str]) -> tuple[int, str, str]:
    """Runs a command and returns (exit code, stdout, stderr) safely."""
    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=15,
            stdin=subprocess.DEVNULL,
        )
        return result.returncode, result.stdout, result.stderr
    except (subprocess.SubprocessError, OSError):
        return -1, "", "comando no disponible"


def _parse_systemctl_units(stdout: str) -> list[ServiceInfo]:
    """Parses `systemctl list-units` output (no legend, no colors)."""
    services = []
    for raw in stdout.splitlines():
        line = raw.strip()
        if not line:
            continue
        # Columns: UNIT LOAD ACTIVE SUB DESCRIPTION
        parts = line.split(None, 4)
        parts += [""] * (5 - len(parts))
        unit, load, active, sub, description = parts[:5]
        if not unit.endswith(".service"):
            continue
        # systemd appends "..." when the description is truncated.
        description = description.rstrip()
        if description.endswith("..."):
            description = description[:-3].rstrip()
        services.append(
            ServiceInfo(
                name=unit,
                description=description,
                load_state=load,
                active_state=active,
                sub_state=sub,
                label=_translate(active),
            )
        )
    return services


def _tally(services: list[ServiceInfo]) -> dict[str, int]:
    counts: dict[str, int] = {}
    for service in services:
        state = service.active_state
        counts[state] = counts.get(state, 0) + 1
    return counts


def _sysv_services() -> list[ServiceInfo]:
    """Lists /etc/init.d scripts (excluding README and hidden files)."""
    init_d = "/etc/init.d"
    if not os.path.isdir(init_d):
        return []
    try:
        names = sorted(os.listdir(init_d))
    except OSError:
        return []
    services = []
    for name in names:
        path = os.path.join(init_d, name)
        if name.startswith(".") or name == "README" or not os.path.isfile(path):
            continue
        services.append(
            ServiceInfo(
                name=name,
                description="Script de inicio SysV",
                load_state="registered",
                active_state="unknown",
                sub_state="unknown",
                label=_translate("registered"),
            )
        )
    return services


def get_services() -> ServicesResponse:
    """Service status with controlled degradation."""
    # --- Preferred: systemd -------------------------------------------------
    if _systemctl_available():
        code, stdout, _stderr = _safe_run(
            [
                "systemctl",
                "list-units",
                "--type=service",
                "--all",
                "--no-pager",
                "--no-legend",
            ]
        )
        if code == 0:
            services = _parse_systemctl_units(stdout)
            if services:
                return ServicesResponse(
                    available=True,
                    manager="systemd",
                    detail="",
                    services=services,
                    counts=_tally(services),
                )
            if _system_is_booted_with_systemd():
                return ServicesResponse(
                    available=True,
                    manager="systemd",
                    detail="systemctl responded but listed no services.",
                    services=[],
                    counts={},
                )

    # --- Fallback: SysV scripts --------------------------------------------
    sysv_services = _sysv_services()
    if sysv_services:
        return ServicesResponse(
            available=True,
            manager="sysv",
            detail="Gestor systemd no arrancado en este entorno; se listan scripts de /etc/init.d.",
            services=sysv_services,
            counts=_tally(sysv_services),
        )

    # --- No accessible manager: controlled response (never crashes) --------
    return ServicesResponse(
        available=False,
        manager=None,
        detail=(
            "No se encontró un gestor de servicios accesible en este entorno "
            "(por ejemplo, WSL sin systemd arrancado). Consulta el README para "
            "activar systemd en WSL."
        ),
        services=[],
        counts={},
    )


def _build_tracked(services: list[ServiceInfo], names: list[str]) -> list[TrackedService]:
    by_name = {service.name: service for service in services}
    tracked = []
    for name in names:
        service = by_name.get(f"{name}.service")
        if service is None:
            tracked.append(
                TrackedService(name=name, state="not_found", label="No instalado", active_state=None)
            )
        elif service.active_state == "failed":
            tracked.append(
                TrackedService(name=name, state="failed", label="Fallido", active_state="failed")
            )
        else:
            tracked.append(
                TrackedService(
                    name=name,
                    state=service.active_state,
                    label=service.label,
                    active_state=service.active_state,
                )
            )
    return tracked


def get_services_with_tracking(tracked_names: list[str]) -> ServicesResponse:
    """Service status plus per-name status for configured tracked services."""
    response = get_services()
    response.tracked = _build_tracked(response.services, tracked_names)
    return response

File: backend-node/src/services/linux-services.ts
import { execFile } from 'child_process';
import * as fs from 'fs';
import { promisify } from 'util';

const execFileAsync = promisify(execFile);

const STATE_LABELS: Record<string, string> = {
  active: 'Activo',
  inactive: 'Inactivo',
  failed: 'Fallido',
  activating: 'Activando',
  deactivating: 'Desactivando',
  exited: 'Finalizado',
  reloading: 'Recargando',
  registered: 'Registrado',
};

const SVC_BINARIES: Record<string, string> = {
  nginx: 'nginx',
  ssh: 'sshd',
  cron: 'cron',
  redis: 'redis-server',
  postgres: 'postgres',
  mysql: 'mysqld',
  mongodb: 'mongod',
};

export interface ServiceInfo {
  name: string;
  description: string;
  load_state: string;
  active_state: string;
  sub_state: string;
  label: string;
}

export interface TrackedService {
  name: string;
  state: string;
  label: string;
  active_state: string | null;
  source: string | null;
  hint: string;
}

export interface ServicesResponse {
  available: boolean;
  manager: string | null;
  detail: string;
  services: ServiceInfo[];
  counts: Record<string, number>;
  tracked: TrackedService[];
}

function translate(state: string): string {
  return STATE_LABELS[state] ?? (state ? state.charAt(0).toUpperCase() + state.slice(1) : 'Desconocido');
}

async function safeRun(cmd: string, args: string[]): Promise<{ code: number; stdout: string; stderr: string }> {
  try {
    const result = await execFileAsync(cmd, args, { timeout: 15000, encoding: 'utf-8', windowsHide: true });
    return { code: 0, stdout: result.stdout, stderr: result.stderr };
  } catch (err) {
    const error = err as { code?: number; stdout?: string; stderr?: string; message?: string };
    return { code: error.code ?? -1, stdout: error.stdout ?? '', stderr: error.stderr ?? error.message ?? 'comando no disponible' };
  }
}

function parseSystemctlUnits(stdout: string): ServiceInfo[] {
  const services: ServiceInfo[] = [];
  for (const raw of stdout.split('\n')) {
    const line = raw.trim();
    if (!line) continue;
    const parts = line.split(/\s+/, 5);
    const unit = parts[0];
    if (!unit || !unit.endsWith('.service')) continue;
    const active = parts[2] ?? '';
    let description = (parts[4] ?? '').trim();
    if (description.endsWith('...')) {
      description = description.slice(0, -3).trim();
    }
    services.push({
      name: unit,
      description,
      load_state: parts[1] ?? '',
      active_state: active,
      sub_state: parts[3] ?? '',
      label: translate(active),
    });
  }
  return services;
}

function tally(services: ServiceInfo[]): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const service of services) {
    const state = service.active_state;
    counts[state] = (counts[state] ?? 0) + 1;
  }
  return counts;
}

function sysvServices(): ServiceInfo[] {
  const initD = '/etc/init.d';
  if (!fs.existsSync(initD)) return [];
  try {
    const names = fs.readdirSync(initD).sort();
    const services: ServiceInfo[] = [];
    for (const name of names) {
      const filePath = `${initD}/${name}`;
      if (name.startsWith('.') || name === 'README') continue;
      if (!fs.statSync(filePath).isFile()) continue;
      services.push({
        name,
        description: 'Script de inicio SysV',
        load_state: 'registered',
        active_state: 'unknown',
        sub_state: 'unknown',
        label: translate('registered'),
      });
    }
    return services;
  } catch {
    return [];
  }
}

function probeDocker(): Promise<{ state: string; label: string; source: string; hint: string }> {
  if (fs.existsSync('/var/run/docker.sock')) {
    return safeRun('docker', ['info', '--format', '{{.ServerVersion}}']).then((result) => {
      if (result.code === 0) {
        return { state: 'active', label: 'Activo', source: 'docker', hint: '' };
      }
      return { state: 'unreachable', label: 'Sin acceso', source: 'docker', hint: 'El motor de contenedores no es accesible desde este entorno.' };
    });
  }
  return safeRun('docker', ['info', '--format', '{{.ServerVersion}}']).then((result) => {
    if (result.code === 0) {
      return { state: 'active', label: 'Activo', source: 'docker', hint: '' };
    }
    if (result.code === -1) {
      return { state: 'not_found', label: 'No instalado', source: 'ninguno', hint: '' };
    }
    return { state: 'unreachable', label: 'Sin acceso', source: 'docker', hint: 'El motor de contenedores no es accesible desde este entorno.' };
  });
}

function probeByProcess(name: string): Promise<{ state: string; label: string; source: string; hint: string }> {
  const binary = SVC_BINARIES[name];
  if (!binary) {
    return Promise.resolve({ state: 'not_found', label: 'No instalado', source: 'ninguno', hint: '' });
  }
  return safeRun('pgrep', ['-x', binary]).then((result) => {
    if (result.code === 0) {
      return { state: 'active', label: 'Activo', source: 'proceso', hint: '' };
    }
    return { state: 'inactive', label: 'Inactivo', source: 'proceso', hint: '' };
  });
}

async function buildTracked(services: ServiceInfo[], names: string[]): Promise<TrackedService[]> {
  const byName: Record<string, ServiceInfo> = {};
  for (const service of services) {
    byName[service.name] = service;
  }
  const tracked: TrackedService[] = [];
  for (const name of names) {
    const service = byName[`${name}.service`];
    if (service) {
      tracked.push({
        name,
        state: service.active_state,
        label: service.label,
        active_state: service.active_state,
        source: 'systemd',
        hint: '',
      });
      continue;
    }
    if (name === 'docker') {
      const probe = await probeDocker();
      tracked.push({
        name,
        state: probe.state,
        label: probe.label,
        active_state: null,
        source: probe.source,
        hint: probe.hint,
      });
      continue;
    }
    const probe = await probeByProcess(name);
    tracked.push({
      name,
      state: probe.state,
      label: probe.label,
      active_state: null,
      source: probe.source,
      hint: '',
    });
  }
  return tracked;
}

export async function getServices(): Promise<ServicesResponse> {
  const hasSystemctl = fs.existsSync('/usr/bin/systemctl') || fs.existsSync('/bin/systemctl');
  if (hasSystemctl) {
    const result = await safeRun('systemctl', ['list-units', '--type=service', '--all', '--no-pager', '--no-legend']);
    if (result.code === 0) {
      const services = parseSystemctlUnits(result.stdout);
      if (services.length > 0) {
        return { available: true, manager: 'systemd', detail: '', services, counts: tally(services), tracked: [] };
      }
      if (process.platform === 'linux') {
        try {
          const comm = fs.readFileSync('/proc/1/comm', 'utf-8').trim();
          if (comm === 'systemd') {
            return { available: true, manager: 'systemd', detail: 'systemctl responded but listed no services.', services: [], counts: {}, tracked: [] };
          }
        } catch {
          // fall through to SysV
        }
      }
    }
  }

  const sysv = sysvServices();
  if (sysv.length > 0) {
    return {
      available: true,
      manager: 'sysv',
      detail: 'Gestor systemd no arrancado en este entorno; se listan scripts de /etc/init.d.',
      services: sysv,
      counts: tally(sysv),
      tracked: [],
    };
  }

  return {
    available: false,
    manager: null,
    detail:
      'No se descubrió un gestor de servicios accesible en este entorno (por ejemplo, WSL sin systemd arrancado). Consulta el README.',
    services: [],
    counts: {},
    tracked: [],
  };
}

export async function getServicesWithTracking(trackedNames: string[]): Promise<ServicesResponse> {
  const response = await getServices();
  response.tracked = await buildTracked(response.services, trackedNames);
  return response;
}